base/rational.jl

```diff
@@ -129,12 +129,299 @@ Bool(x::Rational) = x==0 ? false : x==1 ? true :
+end
+
+rounding_mode_translated_for_abs(
+    rm::Union{
+        RoundingMode{:ToZero},
+        RoundingMode{:FromZero},
+        RoundingMode{:Nearest},
+        RoundingMode{:NearestTiesAway},
+    },
+    ::Real,
+) =
+    rm
+
+rounding_mode_translated_for_abs(::RoundingMode{:Down}, sign::Real) =
+    !signbit(sign) ? RoundToZero : RoundFromZero
+
+rounding_mode_translated_for_abs(::RoundingMode{:Up}, sign::Real) =
+    !signbit(sign) ? RoundFromZero : RoundToZero
+
+# `num`, `den` are positive integers. `requested_bit_width` is the
+# requested floating-point precision. `T` is the integer type that
+# we'll be working with mostly, it needs to be wide enough.
+function rational_to_float_components_impl(
+    num::Integer,
+    den::Integer,
+    requested_bit_width::Integer,
+    ::Type{T},
+    romo::RoundingMode,
+) where {T<:Integer}
+    num_bit_width = bit_width(num)
+    den_bit_width = bit_width(den)
+
+    numT = T(num)
+    denT = T(den)
+
+    # Must be positive.
+    bit_shift_1 = den_bit_width - num_bit_width + requested_bit_width
+    (false ≤ bit_shift_1) || (bit_shift_1 = zero(bit_shift_1))
+
+    # `T` must be wide enough to make overflow impossible during the
+    # left shift here, we must not lose the high bits.
+    #
+    # Similarly, `scaled_num` must not be negative.
+    scaled_num = numT << bit_shift_1
+
+    # `quo0` must be at least `requested_bit_width` bits wide.
+    (quo0, rem0) = divrem(scaled_num, denT, RoundToZero)
+    quo0_bit_width = bit_width(quo0)
+
+    # Now we have a mantissa in `quo0`, but need to round it to the
+    # required precision.
+
+    # Should often be zero, but never negative.
+    bit_shift_2 = quo0_bit_width - requested_bit_width
+
+    # `quo1` is `div(scaled_num, denT << bit_shift_2, RoundToZero)` and should
+    # be exactly `requested_bit_width` bits wide.
+    (quo1, rem1) = divrem_pow2(quo0, bit_shift_2)
+
+    # `rem(scaled_num, denT << bit_shift_2, RoundToZero)`, but without the extra
+    # computation.
+    rem_total = rem1 * den + rem0
+
+    result_is_exact = iszero(rem_total)
+
+    mantissa = quo1
+
+    mantissa_carry = false
+
+    romo_is_to = romo == RoundToZero
+    romo_is_af = romo == RoundFromZero
+    romo_is_ntte = romo == RoundNearest
+    romo_is_ntaf = romo == RoundNearestTiesAway
+
+    romo_is_to_nearest = romo_is_ntte | romo_is_ntaf
+
+    if !result_is_exact & !romo_is_to
+        # Finish the rounding
+
+        (rem_quo, rem_rem) = divrem_pow2(rem_total, bit_shift_2)
+        to_nearest_compar_hi = rem_quo - (den >> true)
+        to_nearest_compar_lo = (rem_rem << true) - ((den & true) << bit_shift_2)
+        to_nearest_compar_hi_iszero = iszero(to_nearest_compar_hi)
+        to_nearest_compar_lo_iszero = iszero(to_nearest_compar_lo)
+        to_nearest_is_greater_than_half =
+            (false < to_nearest_compar_hi) |
+            (
+                to_nearest_compar_hi_iszero &
+                (false < to_nearest_compar_lo)
+            )
```

Comment:
This is really hard to follow - I think that this code is calculating whether the input, when rounded to the nearest floating point, is larger than half of the input or not, but I'm not 100% sure on that. I think written like this is better:

```suggestion
        den_half = den >> one(den)
        twice_rem_rem = rem_rem << one(rem_rem)
        to_nearest_compar_hi = rem_quo == den_half
        to_nearest_compar_lo = twice_rem_rem == (isodd(den) << bit_shift_2)
        to_nearest_is_greater_than_half = to_nearest_compar_hi | to_nearest_compar_lo
```

Fancy tricks to compare booleans numerically seems needlessly convoluted.

I must confess that I don't find variable names like `rem_rem` particularly insightful. What do they represent? "The remainder of a remainder" isn't really a good description of what the variable is for, right?

Reply:
The intention behind making the code unnecessarily convoluted was avoiding calculating `denT << bit_shift_2` to avoid a possible overflow. However, now I did some more experiments and analysis, and the overflow will, in fact, never happen.

So I'll now simplify the code before testing it and pushing again.

Reply:
I did away with the manual rounding code altogether, now the code just does two separate divisions. It can be optimized later if necessary.